asf_heat_pump_affordability/notebooks/identify_and_review_analytical_sample.py

```diff
@@ -0,0 +1,338 @@
+## property archetype
+
+property_archetype_cols = ["CONSTRUCTION_AGE_BAND", "BUILT_FORM", "PROPERTY_TYPE"]
+# df["property_archetype_data_na"] = df[property_archetype_cols].isna().apply(any, axis=1)
+# No need to apply
+df["property_archetype_data_na"] = df[property_archetype_cols].isna().any(axis=1)
+
+## Add col to indicate rows with non-ASHPs
+
+# df["not_ashp"] = df["tech_type"].apply(
+#    lambda x: False if x == "Air Source Heat Pump" else True
+# )
+
+# This is much simpler than the above:
+
+df["not_ashp"] = df["tech_type"] != "Air Source Heat Pump"
+
+# %%
+## Add col to indicate rows which will be dropped based on exclusion criteria
+
+exclusion_criteria = [
+    "cost_na",
+    "tech_type_na",
+    "original_epc_index_na",
+    "property_archetype_data_na",
+    "not_ashp",
+]
+# df["drop_row"] = df[exclusion_criteria].apply(any, axis=1)
+df["drop_row"] = df[exclusion_criteria].any(axis=1)
+
+# %%
+## Plot number of rows removed for each individual exclusion criterion
+
+exclusion_criteria.append("drop_row")
+
+plot_counts = {"labels": exclusion_criteria}
+plot_counts["counts"] = [df[ec].sum() for ec in exclusion_criteria]
+plot_df = pd.DataFrame(plot_counts).sort_values("counts")
+
+fig, ax = plt.subplots()
+bar = ax.bar(x=plot_df["labels"], height=plot_df["counts"])
+ax.bar_label(bar)
+ax.set_ylim([0, round(max(plot_df["counts"]) + 1000, -3)])
+ax.set_title(
+    "Number of rows removed for individual exclusion criteria\nand net exclusion effect"
+)
+plt.xticks(rotation=90)
+plt.show()
+
+# %%
+## See count and proportion of NaN values for each key variable
+## See count and proportion of non-NaN values lost for each key variable when dropping rows based on exclusion criteria
+
+_cols = [
+    "cost_na",
+    "tech_type_na",
+    "original_epc_index_na",
+    "property_archetype_data_na",
+]
+lost_vals = {
+    k: []
+    for k in [
+        "variable",
+        "na_count",
+        "proportion_na",
+        "non_na_lost_count",
+        "non_na_proportion_lost",
+    ]
+}
+
+for _col in _cols:
+    lost_count = df[(df[_col] == False) & (df["drop_row"] == True)]["drop_row"].sum()
+    lost_vals["variable"].append(_col[:-3])
+    lost_vals["na_count"].append(df[_col].sum())
+    lost_vals["proportion_na"].append(df[_col].sum() / len(df))
+    lost_vals["non_na_lost_count"].append(lost_count)
+    lost_vals["non_na_proportion_lost"].append(lost_count / len(df[df[_col] == False]))
+
+pd.DataFrame(lost_vals)
+
+# %%
+# print(f"Total rows dropped: {(df['drop_row'] == True).sum()}")
+# print(f"Total proportion of rows dropped: {(df['drop_row'] == True).sum() / len(df)}")
+
+print(f"Total rows dropped: {df['drop_row'].sum()}")
+print(f"Total proportion of rows dropped: {df['drop_row'].sum() / len(df)}")
+
+# %%
+## Keep rows where drop_row is False, i.e. remove rows based on exclusion criteria
+# sample = df[df["drop_row"] == False]
+sample = df[~df["drop_row"]]
```

Comment:
Sometimes i find the original version with explicit reference easier to read and understand. Is this bad practice or just a matter of preference?

Reply:
Cool I dont feel strongly here so i'll keep it the recommended way. Thanks!